Add morrisjobke/webdav docker container for external storage tests

PVince81 opened the pull request:
Fixes https://github.com/owncloud/core/issues/15897

Note: not all tests pass yet. This is only about introducing the instance.

These are my results:

```
% ./autotest-external.sh sqlite webdav-apache
(...)
Configuration read from /srv/www/htdocs/owncloud/tests/phpunit-autotest-external.xml

............................................F............F..F.... 65 / 71 ( 91%)
......

Time: 1.23 seconds, Memory: 16.50Mb

There were 3 failures:

1) Test\Files\Storage\DAV::testStat
Failed asserting that 100 matches expected 1433252136.

/srv/www/htdocs/owncloud/tests/lib/files/storage/storage.php:303

2) Test\Files\Storage\DAV::testRenameOverWriteDirectory
rename must return true on success
Failed asserting that false is true.

/srv/www/htdocs/owncloud/tests/lib/files/storage/storage.php:462

3) Test\Files\Storage\DAV::testCopyOverWriteDirectory
Failed asserting that true is false.

/srv/www/htdocs/owncloud/tests/lib/files/storage/storage.php:518

FAILURES!
Tests: 71, Assertions: 324, Failures: 3.
stop: stop-webdav-apache.sh
Docker executable found - stop and remove docker containers
Stopping and removing docker container 7c0544b7235e4f976457d27d8f4ce3f31c0d1ba3bf47c3ad380d7f25abe5af3d
7c0544b7235e4f976457d27d8f4ce3f31c0d1ba3bf47c3ad380d7f25abe5af3d
skip: start-webdav-ownCloud.sh

```

Please review @MorrisJobke @DeepDiver1975 @Xenopathic

Commit:
Add morrisjobke/webdav docker container for external storage tests

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/f50cc1c2-ddb5-4ed2-acef-b76567540391) was created.

DeepDiver1975 commented:
Any solution at hand to get rid of the failing tests? We can only set this up on jenkins once we are green - otherwise this will no detect regressions.

Just saying ....

PVince81 commented:
I'll see what I can do

MorrisJobke commented:
@PVince81 Are you sure the WebDAV server works as intended? Maybe it's also a problem with the docker setup. :see_no_evil:

PVince81 commented:
Well some of the fails are about overwriting, maybe Apache's Webdav server doesn't let people overwrite with a MOVE operation like other servers does.

But from your past comments I had the feeling that the docker instance did not work AT ALL (100% fail) but here it seems to work fine for most cases.

PVince81 commented:
but yeah, it could be a setup issue, see for example http://stackoverflow.com/questions/20695683/apache-webdav-unable-to-create-test-folder-when-file-test-txt-exists

PVince81 commented:
Something seems odd with this DAV server:
```
dav:/webdav/> mkdir base
Creating `base': succeeded.
dav:/webdav/> mkdir sub
Creating `sub': succeeded.
dav:/webdav/> mv base sub
Moving `/webdav/base' to `/webdav/sub/base':  redirect to http://localhost:8888/webdav/base/
dav:/webdav/> ls
Listing collection `/webdav/': succeeded.
Coll:   base                                   0  Jul  3 18:58
Coll:   sub                                    0  Jul  3 18:58
dav:/webdav/> 
```

It doesn't properly move.

If you compare with an ownCloud server:
```
dav:/owncloud/remote.php/webdav/> mkdir base
Creating `base': succeeded.
dav:/owncloud/remote.php/webdav/> mkdir sub
Creating `sub': succeeded.
dav:/owncloud/remote.php/webdav/> mv base sub
Moving `/owncloud/remote.php/webdav/base' to `/owncloud/remote.php/webdav/sub/base':  succeeded.
dav:/owncloud/remote.php/webdav/> ls
Listing collection `/owncloud/remote.php/webdav/': succeeded.
Coll:   sub                                    0  Jul  3 19:00
dav:/owncloud/remote.php/webdav/> ls sub
Listing collection `/owncloud/remote.php/webdav/sub/': succeeded.
Coll:   base                                   0  Jul  3 18:59
```

Maybe Apache with mod_dav behaves differently or needs config tweaks.

PVince81 commented:
Yep, it must be the server.
I tried KDE's Dolphin and I also cannot move "base" into "sub" there, getting an error.

PVince81 commented:
I think I found the problem: when doing a Webdav MOVE operation, a "Destination" header must be specified with the **full** URL to the destination, not relative.

Since we are outside the docker instance, the URL is the one of the docker instance, not the one inside. However the Webdav server would only understand it if it was its own internal URL.

PVince81 commented:
Also happening on my VirtualBox VM when connecting to the Webdav server from outside it.
Maybe there is a way to set at alias, maybe some kind of reverse proxy config ?

PVince81 commented:
Strangely renaming files work, maybe Apache is clever enough to find out that it's the same file.

But moving folders doesn't (neither in cadaver not ownCloud).
I tried making the "Destination" field relative but then I get BadRequest. From what I read online it's only allowed in future versions of the Webdav spec...

PVince81 commented:
Okay, it seems that:
- cadaver's MOVE operation doesn't work properly with mod_dav servers with URL differences
- doing a straight curl works: `curl -X MOVE -u test:test "http://test-ubuntuserver2.local:8080/webdav/c/" -H 'Destination: http://test-ubuntuserver2.local:8080/webdav/b/c'`. This explains why all the "move" tests worked
- looks like move + overwrite works too with curl, the condition being that the target path as a trailing slash when it's a directory. Let's try that with our code.

Commit:
Fix webdav destination header when overwriting folders

The trailing slash is needed when talking to Apache's mod_dav server

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/55226a9c-5880-47d2-b66c-444a885c1eb2) was created.

PVince81 commented:
Solved the move/copy + overwrite issue by adding a trailing slash, see https://github.com/owncloud/core/commit/1177969ed79f25507a59cf3e930bdc3152e08b1c

Still, there is the `testStat` that still fails.

PVince81 commented:
```
Configuration read from /srv/www/htdocs/owncloud/tests/phpunit-autotest-external.xml

............................................F.................... 65 / 71 ( 91%)
......

Time: 1.19 seconds, Memory: 16.50Mb

There was 1 failure:

1) Test\Files\Storage\DAV::testStat
Failed asserting that 100 matches expected 1435945545.

/srv/www/htdocs/owncloud/tests/lib/files/storage/storage.php:303

FAILURES!
Tests: 71, Assertions: 329, Failures: 1.
```

PVince81 commented:
Still have trouble with the stat tests

DeepDiver1975 commented:
> Still have trouble with the stat tests

lets move this to 9.0

PVince81 commented:
Looks like either the target DAV server doesn't support "touch" (PROPPATCH with getlastmodified) or we're doing the touch wrong

PVince81 commented:
Usually a proppatch should return a list of updated properties. So we should check whether the "lastmodified" property was actually set, and if not return false.

PVince81 commented:
Indeed, in general Webdav servers do not let you modify "lastmodified": http://stackoverflow.com/a/17979769
ownCloud is the exception.

PVince81 commented:
Looks like Sabre doesn't return the PROPPATCH response, so we cannot find out whether the server accepted the property. Raised here https://github.com/fruux/sabre-dav/issues/726

PVince81 commented:
The next step here is: wait for 8.1.4 to be released because it contains the PROPPATCH fix https://github.com/owncloud/core/pull/19738 which will then appear in the owncloud docker instance. This should at least make the "touch()" operation work properly in that case.

PVince81 commented:
- [ ] REQUIRES: https://github.com/fruux/sabre-dav/issues/726

PVince81 commented:
Looks like the webdav server does accept the "lastmodified" property, but it stores it separately.
"lastmodified" is the correct way to set a mtime... it should be reflected in "getlastmodified" then. But the server doesn't update it (ownCloud does it correctly when acting as DAV server).

Maybe we should read both lastmodified and getlastmodified properties from Webdav and read whichever has a value.

I tried the PROPPATCH getlastmodified instead but the DAV server returns 409 "property is read-only", which is correct...

The problem is: the server accepted the "lastmodified" value as custom property. Ideally we'd like the server to tell us "no, I don't support setting mtimes" so we can switch to our touch fallback... but we can't... and also not sure how to detect this situation.

PVince81 commented:
I wish we could finish this... but... trouble. I tried quickly adding the "lastmodified" property in propfind and co but it fails even harder... something to look into for 9.1

PVince81 commented:
the other alternative would be to skip the "touch" test for webdav so we at least have unit tests for everything else, would that be acceptable @DeepDiver1975 @MorrisJobke ?

PVince81 commented:
It was suggested here that ownCloud could advertise supporting PROPPhttps://github.com/owncloud/core/issues/22499#issue-134549159

If we do this, it would be possible to check that DAV header before attempting a touch. If the server does not advertise that it supports it, then we fallback to the other logic when touch is not supported.

PVince81 commented:
=> discussed with @MorrisJobke and we'll have a look at doing this: after a touch, do a PROPFIND on the target to make sure the mtime was properly accepted. This is needed to find out whether to use the touch fallback.

Commit:
Check whether remote DAV server accepted the mtime on touch

ownCloud as remote DAV always accepts the mtime on touch, but other
servers like Apache's DAV server doesn't. The latter doesn't give any
visible hint in its response to detect this case, so this fix does a
subsequent PROPFIND to check whether the mtime was actually set.

Since a touch() operation seldom happens (only on uploads), the minor
performance loss should hopefully be acceptable.

PVince81 commented:
Done, I added another PROPFIND after PROPPATCH to check if touch succeeded: https://github.com/owncloud/core/pull/16688/commits/a76e1e4a124aac4d396928e9bd72702398c9095d

Please review @MorrisJobke @icewind1991 @nickvergessen @blizzz

MorrisJobke commented on apps/files_external/tests/env/start-webdav-apache.sh:
```diff
@@ -0,0 +1,74 @@
+
+return array(
+    'run'=>true,
+    'host'=>'localhost:$port/webdav/',
```
This will fail on Jenkins. Please use the IP address of the container like in  https://github.com/owncloud/core/blob/f6cea3c9c436142110504ba76320d57ca7899b27/apps/files_external/tests/env/start-sftp-atmoz.sh#L37

MorrisJobke commented:
Beside my above comment this works really nice :+1:

MorrisJobke commented on apps/files_external/tests/env/start-webdav-apache.sh:
```diff
@@ -0,0 +1,74 @@
+#
+# Set environment variable DEBUG to print config file
+#
+# @author Morris Jobke
```
I didn't know that I've written this alone ;)

MorrisJobke commented on apps/files_external/tests/env/start-webdav-apache.sh:
```diff
@@ -0,0 +1,74 @@
+# Set environment variable DEBUG to print config file
+#
+# @author Morris Jobke
+# @copyright 2014 Morris Jobke <hey@morrisjobke.de>
```
Oh nice ... that could be the reason I don't remind this anymore - it's 2 years old ;)

Commit:
Fix webdav test env scripts to also work on CI

PVince81 commented:
Rebased and adjusted env scripts to work on CI

PVince81 commented:
Ah, stupid thing, it hung on the last test. Almost won :frowning: 

Rebasing...

PVince81 commented:
Can you guys review this ? @MorrisJobke @DeepDiver1975 @Xenopathic @icewind1991 

Note that it also comes with some fixes for non-ownCloud webdav

MorrisJobke commented:
Tests looks good 👍

PVince81 commented:
Second review please @icewind1991 @Xenopathic @DeepDiver1975 @rullzer

rullzer commented:
:+1:

PVince81 commented:
We're past feature freeze, however this only adds tests + a bugfix for DAV, so this is acceptable.

Merging.
